src-rocksdb/RocksDB.hs

```diff
@@ -0,0 +1,313 @@
+    rocksdb_readoptions_create
+    rocksdb_readoptions_destroy
+    (\ptr -> kont (ReadOptions ptr))
+
+-------------------------------------------------------------------------------
+-- write options
+-------------------------------------------------------------------------------
+
+newtype WriteOptions = WriteOptions (Ptr WRITEOPTIONS)
+
+withWriteOptions :: (WriteOptions -> IO r) -> IO r
+withWriteOptions kont = bracket
+    rocksdb_writeoptions_create
+    rocksdb_writeoptions_destroy
+    (\ptr -> kont (WriteOptions ptr))
+
+writeOptionsDisableWAL :: WriteOptions -> Bool {- ^ Disable -} -> IO ()
+writeOptionsDisableWAL (WriteOptions opts) disable =
+    rocksdb_writeoptions_disable_WAL opts (if disable then 1 else 0)
+
+-------------------------------------------------------------------------------
+-- db operations
+-------------------------------------------------------------------------------
+
+data RocksDB = RocksDB (Ptr DB) ErrPtr
+
+withRocksDB :: Options -> FilePath -> (RocksDB -> IO r) -> IO r
+withRocksDB (Options opt) path kont =
+    withCString path $ \path' ->
+    withErrPtr $ \errptr ->
+    bracket (rocksdb_open' path' errptr) rocksdb_close $ \ptr ->
+    kont (RocksDB ptr errptr)
+  where
+    rocksdb_open' path' errptr = do
+        ptr <- rocksdb_open opt path' errptr
+        assertErrPtr "rocksdb_open" errptr
+        return ptr
+
+put :: RocksDB -> WriteOptions -> ByteString -> ByteString -> IO ()
+put (RocksDB dbPtr errPtr) (WriteOptions opts) key val =
+    unsafeUseAsCStringLen key $ \(kp, kl) ->
+    unsafeUseAsCStringLen val $ \(vp, vl) -> do
+        rocksdb_put dbPtr opts kp (intToCSize kl) vp (intToCSize vl) errPtr
+        assertErrPtr "rocksdb_put" errPtr
+
+get :: RocksDB -> ReadOptions -> ByteString -> IO (Maybe ByteString)
+get (RocksDB dbPtr errPtr) (ReadOptions opts) key =
+    unsafeUseAsCStringLen key $ \(kp, kl) ->
+    alloca $ \vlPtr -> do
+        vp <- rocksdb_get dbPtr opts kp (intToCSize kl) vlPtr errPtr
+        assertErrPtr "rocksdb_get" errPtr -- TODO: may leak
+        vl <- peek vlPtr
+        if vp == nullPtr
+        then return Nothing
+        else Just <$> unsafePackMallocCStringLen (vp, csizeToInt vl)
+
+multiGet :: RocksDB -> ReadOptions -> [ByteString] -> IO [Maybe ByteString]
+multiGet (RocksDB db _errPtr) (ReadOptions opts) keys =
+    allocaArray n $ \kps ->
+    allocaArray n $ \kls ->
+    allocaArray n $ \vps ->
+    allocaArray n $ \vls ->
+    allocaArray n $ \errs ->
+    withMany unsafeUseAsCStringLen keys $ \keys' -> do
+        -- populate keys
+        ifor_ keys' $ \i (kp, kl) -> do
+            pokeElemOff kps i (constPtr kp)
+            pokeElemOff kls i (intToCSize kl)
+
+        -- multi get
+        rocksdb_multi_get db opts (intToCSize n) kps kls vps vls errs
+
+        -- read keys
+        forM [ 0 .. n - 1 ] $ \i -> do
+            vp <- peekElemOff vps i
+            vl <- peekElemOff vls i
+
+            -- TODO: we don't check errs here. we should.
```

Comment:
Is there anything preventing you from doing so here?

Reply:
Getting them right is something I don't want to spend time on. (Cleaning up resources, it is).